pkg/sql/catalog/tabledesc/mutation.go

```diff
@@ -446,3 +450,28 @@ func newMutationCache(desc *descpb.TableDescriptor) *mutationCache {
 	}
 	return &c
 }
+
+// determineIfIndexNeedsForcePuts based on a given index this function will
+// determine if this mutation should set the force puts flag. This flag indicates
+// that conditional puts are not safe. See index.ForcePut for the exact
+// scenarios.
+func determineIfIndexNeedsForcePuts(index catalog.Index, tableDesc *descpb.TableDescriptor) bool {
+	// If we are doing any of the following mutations, then force puts:
+	//   - delete preserving indexes
+	//   - merging indexes
+	//   - dropping primary indexes
+	if index.Merging() || index.IndexDesc().UseDeletePreservingEncoding ||
+		index.Dropped() && index.IsUnique() && index.GetEncodingType() == catenumpb.PrimaryIndexEncoding {
+		return true
+	}
+
+	if index.GetEncodingType() != catenumpb.PrimaryIndexEncoding {
+		return false
+	}
+	// If we are adding primary indexes with new columns (same key), then
+	// we should also force puts. Attempting conditional puts for UPDATE's
+	// can end badly since we need to compute the expected value with the
+	// new columns.
+	return catalog.MakeTableColSet(tableDesc.PrimaryIndex.KeyColumnIDs...).Difference(
```

Comment:
Expose `Equals` in the `TableColSet` and use it. This will be true for an index which has more columns as the primary key